compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -3280,11 +3280,16 @@ object Types {
           if tp1.hasClassSymbol(defn.NothingClass) then tp2.atoms
           else if tp2.hasClassSymbol(defn.NothingClass) then tp1.atoms
           else tp1.atoms | tp2.atoms
-        val tp1w = tp1.widenSingletons
-        val tp2w = tp2.widenSingletons
-        myWidened = if ((tp1 eq tp1w) && (tp2 eq tp2w)) this else tp1w | tp2w
         atomsRunId = ctx.runId
 
+        myWidened =
+          if isSoft then
+            val tp1w = tp1.widenSingletons
+            val tp2w = tp2.widenSingletons
+            if ((tp1 eq tp1w) && (tp2 eq tp2w)) this else tp1w | tp2w
+          else
+            this
```

Comment:
Similarly this contradicts https://github.com/lampepfl/dotty/blob/a83a0a7e025d4c248291e7832e38699efa794576/compiler/src/dotty/tools/dotc/core/Types.scala#L1229-L1230